### DISCUSSION TITLE
Problem with OsmAnd when using alternative navigation (BRouter)

### DISCUSSION THREAD

**Discussion by henritabs** [osmandapp/OsmAnd#16590] [General]
When using another routing program as the build-in one (car) from OsmAnd, in my case BRouter, neither the speed limit icons / signs nor the lane assist / signs nor the speed camera are shown nor working (no audio on speed limits, speed cameras , etc.). When switching back to build in on everything works again as it supposed to.

Using the build in navigation is unfortunately not an option since calculation of "short" routes (from 100 km) takes eternities (10+ min) and result usually in an app crash. During route calculation Adroid is showing constantly app not responding info.

Brouter is the only alternative I could find that is able to calculate quick and very good routes (compared to Garmin, TomTom and co.) although setup it up is a little fiddling and takes a quite amount of additional space.

Seems to me like a bug and I wasn't able to find a work around. I'm using the latest OsmAnd (4.3.5) and Brouter (1.6.3) version but encountered it also with previous OsmAnd last versions.

Anyone got an idea how to solve this? Thanks.

**Comment by mariush444** [comment]
The same with graphhopper.

My observation is:
When osmand prepares the route then every needed data (speed, camera, bumps, turns) are collected.
Other tools brouter, graphhopper ... just prepare turns and osmand follow up this gpx. (please notice that turns are different than osmand's ones)

I found workaround for speed camera only.
POI overlay -> speed camera - ON  (speed camera POI created as my own in search of categories)
Navigation -> voice prompt -> Nearby POI - ON

_advantage_
every speed camera is mentioned. 
Osmand sometimes doesn't mentione camera when it's definition is set wrong in OSM. e.g. is not attached to the route, highway.
_disadvantage_
every speed camera means false positive cases:
1. camera is on the parallel road or very near on the perpendicular road or on viaduct/bridge that it is under/over you
2. camera measures speed is in opposite direction than you expect

**Reply by henritabs** [reply]
Thanks for the hint. Adding speed cams as POIs works but enabling the voice prompt seems not to work for speed cams. Still other features as lane assistant, allowed speed, etc. are more important at least to me.
It's a pity that this issue seem not to be that important to the developers. OsmAnd could be (one of the) best car navigation and I'm using Garmin, TomTom and GoogleMaps (and tried many other) for several reasons, but if the main feature as route calculation is so bad the rest doesn't matter.

**Reply by mariush444** [reply]
other workaround ...
1. choose profile "online navi" e.g. graphhopper or brouter
2. calculate route to destination
3. when the route is calculated - swipe up the calculated route menu and choose `details`
4. tap `download icon` (top right corner beside the printer icon)
5. `save` route (name of a track can be changed)
6. `dismiss` navigation
7. choose profile car (or any other that's needed)
8. tap navi button then tap options
9. tap `follow track` and choose saved track calculated on-line
10. tap `attach to the route`

PROS
it works as any other route calculated by osmand

CONS
1. I'm not sure if osmand changes or not the calculated route (I tested short distance only. More tests needed)
2. Recalculation can't be done by online/other engine/service.
3. I don't know if during recalculation osmand takes into account route calculating at the begining (in theory yes if `smart route calculation` is on)

**Comment by mikehgentry** [comment]
I have a feeling following the route and using the 'Attach to the roads' option should work, but unfortunately I can't test it because it seems to be broken in the 4.4 beta. Worth a try though

**Reply by mariush444** [reply]
'Attach to the roads'  doesn't help in 4.2.7 

PS
`Adding speed cams as POIs works but enabling the voice prompt seems not to work for speed cams.`
I verified this again and "Nearby POI" works fine in my opinion.

**Comment by mariush444** [comment]
@vshcherb 
I have idea to combine outer route calculation with osmand.
When brouter (offline) or graphhopper (online) calculated the route  (it means we have gpx track) or we have just gpx track we want to follow then we need just the last step made by osmand.
In this case route/track is know and osmand can add just navigation info  (voice/screen alerts) like speed, speed cams ... along the route/track that it is already prepared.

What is your opinion? It seems that the implementation should be quite easy.

**Reply by vshcherb** [reply]
OsmAnd need to attach the route and yes it should be working after that. I'm not sure what is the exact idea cause we already have brouter and online router integration along with gpx navigation

**Reply by mariush444** [reply]
There is integration with brouter and grapphopper and gpx for preparing a route.
So we can choose start point and end point and middle points and the route will be prepared.
The problem is that the route (prepared track) has only information about turns.
There is now information about speed limits along the streets and roads, no speed cams info or tunnels, pedestrian crosswalk and any others voice and screen alerts.

In contrast to above when the route is prepared by osmad routing engine then we have all voice and screen alerts along the route: speed limits info, speed cams info, tunnels, pedestrian crosswalk ... and so on.

Because of that my idea is that after the route was calculated by external engine (brouter, graphhopper, gpx ...) then  we could have possibility to ask osmand to add missing information (like speed limits info, speed cams info, tunnels, pedestrian crosswalk ... and so on) along known route (already prepared route by external engine).

**Reply by vshcherb** [reply]
I think we have that issue to support road attachment for all external routers, we already support it for Generic GPX navigation which is possible to configure as generic url

**Reply by mariush444** [reply]
I'm sorry, I didn't catch. What do you mean "configure as generic **_url_**"?

**Reply by vshcherb** [reply]
There is Online Router -> GPX  - I mean you could try it

**Reply by henritabs** [reply]
@vshcherb How / where can I create the request for road attachments for extertal routers?
That would make Osmand to usefull car navigation. I'm still suprised that the native router is not able to calculate routes of 100km or longer. As mentioned before I still think Osm could be with it's features one or the best navigation but navigation is the main feature and has to work :-).
